error_handling.md

```diff
@@ -0,0 +1,88 @@
+```v
+fn foo() string | none | Error {...}
+```
+whereas the part `string | none` can be shortened with `?string`.
+It would result in
+```v
+fn foo() ?string | Error {...}
+```
+where `?` is a shortcut for a sum type `Option` whose implementation could look like this:
+```
+type Option<T> = T | none
+``` 
+Error wouldn't have a syntax sugar in this case.
+
+### Downsides of this specification
+- Anonymous sum types are not yet implemented and needs to be implemented first
+- Generic sum types are not yet implemented and needs to be implemented first
+- The Option sum type needs to be compiler generated rather than syntax generated to make it possible to be used inside `builtin`. Caching generics isn't possible yet.
+- `or{}` block would be the Jack of all trades because it is used for two special kinds of sum types that are not related to each other and doesn't share anything.
+- It would be possible to define variables like `x := Option<string>('foo') // or ?string('foo')` (it's obviously a sum type that can be used everywhere) but this feature should only be used for return types of functions and struct fields. One solution would be to restrict this sum type.
+
+### Or interface
+Since both, `Option` and `Result` would be handled by an `or{}` block, both types could implement a compiler internal interface `Or` that implements the syntax sugar `or {...}`.
+
+### Syntax
+In most languages the questionmark `?` stands for optional values and could be used for `Option`. The equivalent operator would be the excalamtion mark `!` that could be used for `Result`.
+That means `?string` would mean `Option<string>` and `!string` would mean `Result<string>`. Having the syntax of `?` and `!` would help indicating that's a compiler feature and it wouldn't be needed to use sum types that obviously has some overhead.
+
+There is only one tricky part: `?!string` or `!?string`. What would it mean? Which ordner is needed? One solution would be to define that the order doesn't matter.
+`!?` or `?!` always results in `Result<Option<T>>` or `T | none | Error`. But V fmt could always reformat it to `!?` for example.
+> This is even a rare case. It's complexity isn't that awful.
+
+### Handling `Result<Option<T>>`
+Handling `?T` and `!T` is very easy and it will stay as it currently is:
+```v
+foo() or {...}
+```
+for both cases. But the `!T` adds an `err` variable to the `or{}` context. The downside of this is that the user would need to know which result arrives and he is dependent on a language sever to know it.
+An alternative approach would be `or{}` for `Option` and `catch{}` for `Result`. A catch block always has the `err` variable and the `or{}` block nerver has it.
+
+Handling the `!?T`/`Result<Option<T>>` with `or{}` would result in the initial behavior, we already have. For this a solution would be to extend the functionality of `match`.
+Matching the `!?T` type would always look like this:
```

Comment:
that is super confusing

Reply:
`T | IError | None` instead of `Result<Option<T>>` or `!?T` or `?!T` seems way clearer to me

Reply:
Why is it confusing to you? The reason I did not suggest it here is because `?string` is not a sum type but `T | IError | None` is a sum type. I wanted to avoid sum type syntax to be clean and tell the user it works different.
Since it's a rare case, why does it matter that much?

Reply:
Because you will always wonder whether it is `!?T` or `?!T` and whether the order matters. With an anonymous sumtype, there is no confusion, and what is more, it is immediately clear that it must be handled in a `match` or with smartcasts, and NOT with an `or {}` block (which can only handle 2 variants, not 3, without confusion)

Reply:
Since we discussed this in discord i will resolve this and update the spec.